pkg/ingester/stream.go

```diff
@@ -209,96 +298,69 @@ func (s *stream) Push(
+		if !isReplay && s.unorderedWrites && !highestTs.IsZero() && cutoff.After(entries[i].Timestamp) {
 			failedEntriesWithError = append(failedEntriesWithError, entryWithError{&entries[i], chunkenc.ErrTooFarBehind(cutoff)})
 			outOfOrderSamples++
 			outOfOrderBytes += len(entries[i].Line)
-		} else if err := chunk.chunk.Append(&entries[i]); err != nil {
-			failedEntriesWithError = append(failedEntriesWithError, entryWithError{&entries[i], err})
-			if chunkenc.IsOutOfOrderErr(err) {
-				outOfOrderSamples++
-				outOfOrderBytes += len(entries[i].Line)
-			}
-		} else {
-			storedEntries = append(storedEntries, entries[i])
-			s.lastLine.ts = entries[i].Timestamp
-			s.lastLine.content = entries[i].Line
-			if s.highestTs.Before(entries[i].Timestamp) {
-				s.highestTs = entries[i].Timestamp
-			}
-			s.entryCt++
+			continue
+		}
+
+		totalBytes += len(entries[i].Line)
 
-			// length of string plus
-			bytesAdded += len(entries[i].Line)
+		lastLine.ts = entries[i].Timestamp
+		lastLine.content = entries[i].Line
+		if highestTs.Before(entries[i].Timestamp) {
+			highestTs = entries[i].Timestamp
 		}
-		chunk.lastUpdated = time.Now()
+
+		toStore = append(toStore, entries[i])
 	}
 
-	if len(storedEntries) != 0 {
-		// record will be nil when replaying the wal (we don't want to rewrite wal entries as we replay them).
-		if record != nil {
-			record.AddEntries(uint64(s.fp), s.entryCt, storedEntries...)
-		} else {
-			// If record is nil, this is a WAL recovery.
-			s.metrics.recoveredEntriesTotal.Add(float64(len(storedEntries)))
+	// Each successful call to 'AllowN' advances the limiter. With all-or-nothing
+	// ingestion, the limiter should only be advanced when the whole stream can be
+	// sent
+	now := time.Now()
+	if s.cfg.RateLimitWholeStream && !s.limiter.AllowN(now, totalBytes) {
+		// Report that the whole stream was rate limited
+		rateLimitedSamples = len(entries)
+		failedEntriesWithError = make([]entryWithError, 0, len(entries))
+		for i := 0; i < len(entries); i++ {
+			failedEntriesWithError = append(failedEntriesWithError, entryWithError{&entries[i], &validation.ErrStreamRateLimit{RateLimit: flagext.ByteSize(limit), Labels: s.labelsString, Bytes: flagext.ByteSize(len(entries[i].Line))}})
+			rateLimitedBytes += len(entries[i].Line)
 		}
+	}
 
-		s.tailerMtx.RLock()
-		hasTailers := len(s.tailers) != 0
-		s.tailerMtx.RUnlock()
-		if hasTailers {
-			go func() {
-				stream := logproto.Stream{Labels: s.labelsString, Entries: storedEntries}
-
-				closedTailers := []uint32{}
-
-				s.tailerMtx.RLock()
-				for _, tailer := range s.tailers {
-					if tailer.isClosed() {
-						closedTailers = append(closedTailers, tailer.getID())
-						continue
-					}
-					tailer.send(stream, s.labels)
-				}
-				s.tailerMtx.RUnlock()
-
-				if len(closedTailers) != 0 {
-					s.tailerMtx.Lock()
-					defer s.tailerMtx.Unlock()
+	s.reportMetrics(outOfOrderSamples, outOfOrderBytes, rateLimitedSamples, rateLimitedBytes)
```

Comment:
Hmm double checking: I believe you are discarding out of order-related metrics two times: here and in the storeEntries func.

Reply:
Things won't be counted twice because we won't try to store anything filtered by `validation`.